File: learning-one-neuron.js
var problem = [
    '...?.....?............#',
    '......?.......?....####',
    '.?........?.....#####?#',
    '.......?.....#####?####',
    '..?..?....#############',
    '.......######?######?##',
    '....#####?#######?#####',
   '.###################?##'
];

var wx = 0;
var wy = 0;
var wb = 0;

function neuron(x, y, bias=1) {
    let f = wx * x + wy * y + wb * bias;
    if (f < 0) {
        return -1;
	}
    return 1;
}

function learn(x, y, desired) {
    let learning_constant = 0.1;
    let error = desired - neuron(x, y);
    wx += error * x * learning_constant;
    wy += error * y * learning_constant;
    wb += error * 1 * learning_constant;  // bias = 1
}

function one_step_of_learning() {
    let y = 0;
    for (let line of problem) {
        let x = 0;
        for (let char_ of line) {
            if (char_ === '.')
                learn(x, y, -1);
            if (char_ === '#')
                learn(x, y, 1);
            ++x;
		}
        ++y;
	}
}

function print_solution() {
    let y = 0;
    for (let line of problem) {
        let x = 0;
        let new_line = "";
        for (let char_ of line) {
            if (char_ == '?') {
                let r = neuron(x, y);
                if (r < 0)
                    char_ = '!';
                else
                    char_ = '%';
			}
            new_line += char_;
            ++x;
		}
        console.log(new_line);
        ++y;
	}
}

var y = 0;
for (let line of problem) {
    let x = 0;
    let new_line = "";
    for (let char_ of line) {
        if (char_ == '?') {
            let r = neuron(x, y);
            if (r < 0)
                char_ = '!';
            else
                char_ = '%';
		}
        new_line += char_;
        x += 1;
	}
    console.log(new_line);
    y += 1;
}

print_solution();
for (let i = 0; i < 100; ++i) {
    one_step_of_learning();
    print_solution();
}

File: learning_one_neuron.lua

local problem = {
   "...?.....?............#",
   "......?.......?....####",
   ".?........?.....#####?#",
   ".......?.....#####?####",
   "..?..?....#############",
   ".......######?######?##",
   "....#####?#######?#####",
   ".###################?##"}

local wx = 0
local wy = 0
local wb = 0


function neuron(x, y, bias)
   bias = bias or 1
   local f = wx * x + wy * y + wb * bias  -- local f = wx * (x-1) + wy * (y-1) + wb * bias
   if (f < 0) then
      return -1;
   end
   return 1; 
end


function learn(x, y, desired)
   local learning_constant = 0.1
   local _error = desired - neuron(x, y)
   wx = wx + _error * x * learning_constant  -- wx = wx + _error * (x-1) * learning_constant
   wy = wy + _error * y * learning_constant  -- wy = wy + _error * (y-1) * learning_constant
   wb = wb + _error * 1 * learning_constant  -- bias = 1
end


function one_step_of_learning()
   for y = 1, #problem do  -- for y, line in ipairs(problem)
      for x = 1, #problem[y] do  -- string.len(problem[y])  -- for x, char in ipairs(line)
	     local char_ = problem[y]:sub(x, x)
		 if (char_ == '.') then
		    learn(x, y, -1)
		 end
		 if (char_ == '#') then
		    learn(x, y, 1)
		 end
	  end
   end
end


function print_solution()
   for y = 1, #problem do  -- for y, line in ipairs(problem)
      local new_line = ""
      for x = 1, #problem[y] do  -- string.len(problem[y])  -- for x, char in ipairs(line)
	     local char_ = problem[y]:sub(x, x)
         if (char_ == '?') then
		    local r = neuron(x, y)
			if (r < 0) then
			   char_ = '!'
            else
               char_ = '%'
            end
		 end
		 new_line = new_line ..char_
	  end
	  print(new_line)
   end
end


print_solution()
for i = 1, 100 do
   one_step_of_learning()
   print_solution()
end
